FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -696,11 +759,189 @@ static MQTTStatus_t prvMQTTConnect( MQTTContext_t * pxMQTTContext,
     LogInfo( ( "Session present: %d", xSessionPresent ) );
     configASSERT( xResult == MQTTSuccess );
 
+    /* Resend publishes if session is present. NOTE: It's possible that some
+     * of the operations that were in progress during the network interruption
+     * were subscribes. In that case, we would want to mark those operations
+     * as completing with error and remove them from the list of operations, so
+     * that the calling task can try subscribing again. We do not handle that
+     * case in this demo for simplicity, since only one subscription packet is
+     * sent per iteration of this demo. */
+    if( xSessionPresent )
+    {
+        MQTTStateCursor_t cursor = MQTT_STATE_CURSOR_INITIALIZER;
+        uint16_t packetId = MQTT_PACKET_ID_INVALID;
+        AckInfo_t xFoundAck;
+
+        packetId = MQTT_PublishToResend( &globalMqttContext, &cursor );
+
+        while( packetId != MQTT_PACKET_ID_INVALID )
+        {
+            /* Retrieve the operation but do not remove it from the list. */
+            xFoundAck = prvGetAwaitingOperation( packetId, false );
+
+            if( xFoundAck.usPacketId == packetId )
+            {
+                /* Set the DUP flag. */
+                xFoundAck.xOriginalCommand.pxCmdContext->pxPublishInfo->dup = true;
+                xResult = MQTT_Publish( &globalMqttContext, xFoundAck.xOriginalCommand.pxCmdContext->pxPublishInfo, packetId );
+            }
+
+            packetId = MQTT_PublishToResend( &globalMqttContext, &cursor );
```

Comment:
Nit: As there are two calls to `MQTT_PublishToResend` with the same parameters, would it be possible to use a do...while instead?

Reply:
I don't think a do-while can be used since the publish should execute 0 or more times, not 1 or more, and I think adding an extra `if` check for a 0 packet ID looks worse than just keeping this.